api/handlers.go

```diff
@@ -656,6 +657,43 @@ func (api *API) userPUT(u *database.User, w http.ResponseWriter, req *http.Reque
 	api.loginUser(w, u, true)
 }
 
+// userPubKeyDELETE removes a given pubkey from the list of pubkeys associated
+// with this user. It does not require a challenge-response because the used
+// does not need to prove the key is theirs.
+func (api *API) userPubKeyDELETE(u *database.User, w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	ctx := req.Context()
+	var pk database.PubKey
+	err := pk.LoadString(ps.ByName("pubKey"))
+	if err != nil {
+		api.WriteError(w, err, http.StatusBadRequest)
+		return
+	}
+	if !u.HasKey(pk) {
+		// This pubkey does not belong to this user.
+		api.WriteError(w, errors.New("the given pubkey is not associated with this user"), http.StatusBadRequest)
+		return
+	}
+	// Find the position of the pubkey in the list.
+	keyIdx := -1
+	for i, k := range u.PubKeys {
+		if subtle.ConstantTimeCompare(pk[:], k[:]) == 1 {
+			keyIdx = i
+			break
+		}
+	}
+	if keyIdx == -1 {
+		build.Critical("Reaching this should be impossible. It would indicate a concurrent change of the user struct.")
+	}
+	// Remove the pubkey.
+	u.PubKeys = append(u.PubKeys[:keyIdx], u.PubKeys[keyIdx+1:]...)
+	err = api.staticDB.UserSave(ctx, u)
```

Comment:
It might be safer to create a dedicated pubkey deletion method which uses [$pull](https://docs.mongodb.com/manual/reference/operator/update/pull/) here. Because then you can call this method on 2 different servers with different pubkeys without running into a race condition. This way, one call might add back the pubkey of the previous one if the timing was unfortunate. 

In a perfect world (probably not worth it right now) this would be the only db operation happening when this endpoint is called. Including the authentication.

Reply:
Oh, you're right, this is potentially very dangerous.

Reply:
So, this is now tackled but the solution is a little incomplete.

What is done:
When we're adding pubkeys we check if the list of pubkeys is `null`. If it is, we save the entire user record in order to create the array of pubkeys. Otherwise, we use `$addToSet` or `$pull`.

The issue is that we cannot use `$addToSet` when the array is not initialised in mongo - it fails with `Cannot apply $addToSet to non-array field. Field named 'pub_keys' has non-array type null` and there doesn't seem to be a nice way around it.

I'll mitigate this in a follow up by locking the user record any time I change it. I need to do this for other reasons as well. I don't want to do it in this PR because it will blow it up.